### PR TITLE
Helidon common context One pager.

### PR DESCRIPTION
**Pull request by tomas-langer**
Ope pager with proposal of common context handling in Helidon.

The "why" is described in the one pager.

Signed-off-by: Tomas Langer <tomas.langer@oracle.com>

### CODE REVIEW

**Comment by tomas-langer** [comment]
See #600 for first reference

**Comment by ljnelson** [comment]
Do I understand right this is basically CDI scopes (outside of CDI obviously)?

**Comment by tomas-langer** [comment]
This is similar to CDI request scope. The Context is actually hierarchical - the context can have a parent.
So you have a single point to access context related information (which may come from request specific context, or may be a part of the overall application (e.g. webserver) - such as `Tracer`)

File: docs-internal/context.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                            .accept(nextRequest, nextResponse));
+```
+
+Processing through an executor service (`JerseySupport`):
+```java
+ExecutorService executorService = (service != null) 
+                ? service 
+                : Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() * 2);
+        
+this.service = Contexts.wrap(executorService);
```
Comment (batsatt) [inline docs-internal/context.md]: I assume this will do nothing if `executorService` is already wrapped?

Comment (tomas-langer) [reply docs-internal/context.md]: Yes, I have already pushed that fix

Comment (tomas-langer) [reply docs-internal/context.md]: Did not realize this is the one pager PR. You can have a look on the first impl in #600

**Comment by tomas-langer** [comment]
@romain-grecourt @spericas can you please add your opinion? I would like to have at least two approvals before merging this and continuing with the implementation

**Comment by tjquinno** [comment]
I know I am very late to this, and it largely looks very good and very useful.

A few quick things, and I apologize if these have been discussed elsewhere already.

Does `inContext` automatically clear the context in the thread used for the `Runnable` or `Callable` after that code has finished or failed due to an exception? 

Is there need for an explicit `clear` operation to disassociate a context from a thread?

@batsatt asked earlier about preventing multiple wrapping for executor services. 

Similarly, does `Contexts.inContext` _replace_ any existing context that might be present or does it `wrap` it, creating nested contexts? This could presumably happen if the `Runnable` or `Callable` passed to `inContext` itself invoked `inContext`. 

If nested contexts can exist, what does that imply about what the "current" context is...the lowest-level, some kind of merge of the nested contexts? 

On the other hand, if `inContext` _replaces_ any current context then is that something we want to allow...put another way, is that a bug? Should we enforce that a thread's context be `clear``ed before being overridden by another `inContext` invocation?

Does each `Context` have a unique identifier, probably most useful for logging and debugging.

Some of these might not make sense in the use cases we _intend_ now, but a general and powerful mechanism like this might be used in clever ways in the future that we don't anticipate now.

**Comment by tomas-langer** [comment]
The `Context.inContext` is intended to clean up after itself.
If you look to the first implementation, I am using a stack.
The `inContext` method is implemented as follows:
```
public static void inContext(Context context, Runnable runnable) {
        push(context);
        try {
            runnable.run();
        } finally {
            pop();
        }
    }
```

**Comment by tomas-langer** [comment]
The `clear` operation would require quite a big change, as the stack approach would fail in case you clear the context. 
As the implementation now uses a `finally` block, the context will get cleared correctly.
The hierarchy is intentional - you may want to create a child context and run further processing in the child context, rather than the original (parent) context.
The unique identifier sounds good, I will check the implementation and add it if not there.

**Comment by tomas-langer** [comment]
Related to #596

**Comment by tomas-langer** [comment]
I am merging this in. Let's discuss further comments on the implementation

**Pull request merged**